### PR TITLE
adds ceph-volume support

### PR DESCRIPTION
**Pull request by deep-j**
This commit
-adds default value of 10GB to bluestore_block_size in the config file
-adds a check for osd device, upon block device it reads bluestoreblocksize environment variable if available and sets appropriate block size in the config file and uses ceph-volume for deployment.
Fixes: #1

### CODE REVIEW

**Comment by leseb** [comment]
Please rebase your branch.

File: cmd/init_osd.go [hunk 1]
```diff
@@ -43,22 +44,78 @@ func bootstrapOsd() {
 	monDataPath := cephDataPath + "/mon/ceph-" + hostname
 	monKeyringPath := monDataPath + "/keyring"
 
+	// check for block device
+        log.Println("init osd: checking for block device")
```
Comment (leseb) [inline cmd/init_osd.go]: Wrong indent

File: cmd/init_osd.go [hunk 2]
```diff
@@ -43,22 +44,78 @@ func bootstrapOsd() {
...
+	osdDeviceEnv := os.Getenv("OSD_DEVICE")
+	bluestoreBlockSizeEnv := os.Getenv("BLUESTORE_BLOCK_SIZE")
+	if len(osdDeviceEnv)>0 {
```
Comment (leseb) [inline cmd/init_osd.go]:
```suggestion
	if len(osdDeviceEnv) > 0 {
```

File: cmd/init_osd.go [hunk 3]
```diff
@@ -43,22 +44,78 @@ func bootstrapOsd() {
...
+		testDev, err := getFileType(osdDeviceEnv)
+		if err != nil {
+			log.Fatal(err)
+		}
+		log.Println("osd: check for blockdev")
```
Comment (leseb) [inline cmd/init_osd.go]:
```suggestion
		log.Println("init osd: check for blockdev")
```

File: cmd/init_osd.go [hunk 4]
```diff
@@ -43,22 +44,78 @@ func bootstrapOsd() {
...
+		if testDev == "blockdev" {
+			if len(bluestoreBlockSizeEnv)>0 {
```
Comment (leseb) [inline cmd/init_osd.go]:
```suggestion
			if len(bluestoreBlockSizeEnv) > 0 {
```

**Comment by leseb** [comment]
@deep-j is it still WIP?

File: cmd/init_osd.go [hunk 5]
```diff
@@ -43,22 +44,104 @@ func bootstrapOsd() {
...
+		if err != nil {
+			log.Fatal(err)
+		}
+		if testDev == "blockdev" {
+			if len(bluestoreBlockSizeEnv) > 0 {
+				size := toBytes(bluestoreBlockSizeEnv)
+				// override the default value with actual size of block device available
+				sedFile(cephConfFilePath, "bluestore_block_size = 10737418240", "bluestore_block_size = "+string(size))
+			}
+
```
Comment (leseb) [inline cmd/init_osd.go]: remove empty line

File: cmd/init_osd.go [hunk 6]
```diff
@@ -43,22 +44,104 @@ func bootstrapOsd() {
...
+			cmd = exec.Command("ceph-volume", "lvm", "prepare", "--data", osdDeviceEnv)
+
+			out, err = cmd.CombinedOutput()
+			if err != nil {
+				fmt.Printf("The command was: %s\n", cmd.Args)
+				fmt.Printf("The error was: %s\n", out)
+				log.Fatal(err)
+			}
+
```
Comment (leseb) [inline cmd/init_osd.go]: remove empty line

File: cmd/init_osd.go [hunk 7]
```diff
@@ -43,22 +44,104 @@ func bootstrapOsd() {
...
+			// chown osd keyring
+			err = os.Chown(osdKeyringPath, cephUID, cephGID)
+			if err != nil {
+				log.Fatal(err)
+			}
+			// populate osd store
+			osdMkfs(monInitialKeyringPath)
+		}
+	}
```
Comment (leseb) [inline cmd/init_osd.go]: add empty line

File: cmd/init_osd.go [hunk 8]
```diff
@@ -43,22 +44,104 @@ func bootstrapOsd() {
...
+
+				out, err := cmd.CombinedOutput()
+				if err != nil {
+					fmt.Printf("The command was: %s\n", cmd.Args)
+					fmt.Printf("The error was: %s\n", out)
+					log.Fatal(err)
+				}
+
+			} else {
+				log.Fatal("Error occurred osd fsid is not a string")
```
Comment (leseb) [inline cmd/init_osd.go]: This error is meaningless for humans.

Comment (deep-j) [reply cmd/init_osd.go]: Should the error message be removed entirely or changed to "Could not initiate block device activation ".

**Comment by deep-j** [comment]
> @deep-j is it still WIP?


Currently the bluestore_block_size is updated if $BLUESTORE_BLOCK_SIZE is set. There is no scan of block device done to get its actual size.

File: cmd/init_osd.go (hunk 8 above)
Comment (leseb) [reply cmd/init_osd.go]: Change to something meaningful.

**Comment by deep-j** [comment]
> > @deep-j is it still WIP?
> 
> Currently the bluestore_block_size is updated if $BLUESTORE_BLOCK_SIZE is set. There is no scan of block device done to get its actual size.

Added scan of block device when $BLUESTORE_BLOCK_SIZE is not available.

**Comment by leseb** [comment]
@deep-j can you run `cn-core` from inside a container and then get the ceph status?

File: cmd/init_mon.go [hunk 9]
```diff
@@ -52,6 +52,7 @@ osd objectstore = bluestore
 osd memory target = %d
 osd memory base = %d
 osd memory cache min = %d
+bluestore_block_size = 10737418240
```
[the remark this answers is not in the document]
Comment (deep-j) [reply cmd/init_mon.go]: Added a check.
https://github.com/ceph/cn-core/blob/2e26370eb518e9ecd844a3692887e055a67944e6/cmd/init.go#L71-L78

File: cmd/init_mon.go [hunk 10]
```diff
@@ -40,7 +40,7 @@ const (
 [global]
 fsid = %s
 mon initial members = %s
-mon host = v2:127.0.0.1:3300/0
+mon host = 127.0.0.1
```
Comment (leseb) [inline cmd/init_mon.go]: Make a separate commit for this.

Comment (deep-j) [reply cmd/init_mon.go]: Updated

Comment (leseb) [inline cmd/init_mon.go]: Actually, the syntax is: `[v2:127.0.0.1:3300,v1:127.0.0.1:6789]`. Let's try it.

Comment (deep-j) [reply cmd/init_mon.go]: Updated
